### ISSUE TITLE
False positive `arc_with_non_send_sync` on generic Mutex

### ISSUE THREAD

**Issue by kpreid** [rust-lang/rust-clippy#11090]
### Summary

This is a false positive since `M` _may be_ `Send + Sync`.

### Lint Name

arc_with_non_send_sync

### Reproducer

I tried this code:

```rust
use std::sync::{Arc, RwLock};

pub struct Sink<M> {
    messages: Arc<RwLock<Vec<M>>>,
}

impl<M> Sink<M> {
    pub fn new() -> Self {
        Self {
            messages: Arc::new(RwLock::new(Vec::new())),
        }
    }
}
```

I saw this happen:

```
error: usage of `Arc<T>` where `T` is not `Send` or `Sync`
  --> src/lib.rs:10:23
   |
10 |             messages: Arc::new(RwLock::new(Vec::new())),
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: consider using `Rc<T>` instead or wrapping `T` in a std::sync type like `Mutex<T>`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arc_with_non_send_sync
   = note: `#[deny(clippy::arc_with_non_send_sync)]` on by default
```

### Version

```text
rustc 1.72.0-nightly (839e9a6e1 2023-07-02)
binary: rustc
commit-hash: 839e9a6e1210934fd24b15548b811a97c77138fc
commit-date: 2023-07-02
host: x86_64-apple-darwin
release: 1.72.0-nightly
LLVM version: 16.0.5
```

**Comment by kpreid** [comment]
Whoops, I started this after searching last night and it's a duplicate of #11076.
